cairo_programs/sha256.cairo

```diff
@@ -0,0 +1,307 @@
+    tempvar range_check_ptr = range_check_ptr
+    tempvar m = SHA256_STATE_SIZE_FELTS
+
+    output_loop:
+    tempvar x0 = sha256_ptr[0 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr] = x0
+    assert [range_check_ptr + 1] = MAX_VALUE - x0
+    tempvar x1 = sha256_ptr[1 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 2] = x1
+    assert [range_check_ptr + 3] = MAX_VALUE - x1
+    tempvar x2 = sha256_ptr[2 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 4] = x2
+    assert [range_check_ptr + 5] = MAX_VALUE - x2
+    tempvar x3 = sha256_ptr[3 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 6] = x3
+    assert [range_check_ptr + 7] = MAX_VALUE - x3
+    tempvar x4 = sha256_ptr[4 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 8] = x4
+    assert [range_check_ptr + 9] = MAX_VALUE - x4
+    tempvar x5 = sha256_ptr[5 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 10] = x5
+    assert [range_check_ptr + 11] = MAX_VALUE - x5
+    tempvar x6 = sha256_ptr[6 * SHA256_INSTANCE_SIZE]
+    assert [range_check_ptr + 12] = x6
+    assert [range_check_ptr + 13] = MAX_VALUE - x6
+    assert outputs[0] = x0 + 2 ** 35 * x1 + 2 ** (35 * 2) * x2 + 2 ** (35 * 3) * x3 +
+        2 ** (35 * 4) * x4 + 2 ** (35 * 5) * x5 + 2 ** (35 * 6) * x6
+
+    tempvar outputs = outputs + 1
+    tempvar sha256_ptr = sha256_ptr + 1
+    tempvar range_check_ptr = range_check_ptr + 14
+    tempvar m = m - 1
+    jmp output_loop if m != 0
+
+    return _finalize_sha256_inner(
+        sha256_ptr=sha256_start + SHA256_INSTANCE_SIZE * BLOCK_SIZE,
+        n=n - 1,
+        round_constants=round_constants)
+end
+
+func finalize_sha256{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(
+        sha256_ptr_start : felt*, sha256_ptr_end : felt*):
+    alloc_locals
+
+    let (__fp__, _) = get_fp_and_pc()
+
+    let (round_constants) = get_round_constants()
+
+    tempvar n = (sha256_ptr_end - sha256_ptr_start) / SHA256_INSTANCE_SIZE
+    if n == 0:
+        return ()
+    end
+
+    %{
+        # Add dummy pairs of input and output.
+        from starkware.cairo.common.cairo_sha256.sha256_utils import (
+            IV, compute_message_schedule, sha2_compress_function)
+
+        _block_size = int(ids.BLOCK_SIZE)
+        assert 0 <= _block_size < 20
+        _sha256_input_chunk_size_felts = int(ids.SHA256_INPUT_CHUNK_SIZE_FELTS)
+        assert 0 <= _sha256_input_chunk_size_felts < 100
+
+        message = [0] * _sha256_input_chunk_size_felts
+        w = compute_message_schedule(message)
+        output = sha2_compress_function(IV, w)
+        padding = (message + IV + output) * (_block_size - 1)
+        segments.write_arg(ids.sha256_ptr_end, padding)
+    %}
+
+    # Compute the amount of blocks (rounded up).
+    let (local q, r) = unsigned_div_rem(n + BLOCK_SIZE - 1, BLOCK_SIZE)
+    _finalize_sha256_inner(sha256_ptr_start, n=q, round_constants=round_constants)
+    return ()
+end
+
+func main{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}():
```

Comment:
Maybe it would be better to make another Cairo program that imports the necessary functions and performs this computation for the test. What do you think?

Reply:
Resolved:
https://github.com/lambdaclass/cleopatra_cairo/pull/357/commits/482bab53609b9122129e345fc0288b96eede6a9a